### PR TITLE
epic: Blob 4844 DA

### PR DESCRIPTION
**Pull request by icemelon**
### Description

EIP-4844 Blob DA

### CODE REVIEW

**Comment by kunxian-xia** [comment]
PRs need to be merged before the integration test passes: 
- [x] #1155 
- [x] #1160

File: zkevm-circuits/src/pi_circuit.rs [hunk 1]
```diff
@@ -60,12 +61,7 @@ use halo2_proofs::{circuit::SimpleFloorPlanner, plonk::Circuit};
 use itertools::Itertools;
 
 fn get_coinbase_constant() -> Address {
-    let default_coinbase = if cfg!(feature = "scroll") {
-        Address::from_str("0x5300000000000000000000000000000000000005").unwrap()
-    } else {
-        Address::zero()
-    };
-    read_env_var("COINBASE", default_coinbase)
+    read_env_var("COINBASE", Address::zero())
```
Comment (lispc) [inline zkevm-circuits/src/pi_circuit.rs]: @darth-cy why this change?

File: zkevm-circuits/src/tx_circuit.rs [hunk 2]
```diff
@@ -2367,7 +2600,30 @@ impl<F: Field> TxCircuitConfig<F> {
...
+
+        let hash_len = if is_chunk_bytes {
+            tx.rlp_signed.len()
+        } else {
+            0
+        };
+        let tx_hash_rlc = rlc_be_bytes(&tx.rlp_signed, keccak_input);
+        if hash_len >= pows_of_rand.len() {
+            for _ in 0..(tx.rlp_signed.len() - pows_of_rand.len() + 1) {
+                pows_of_rand.push(pows_of_rand.last().unwrap().mul(keccak_input));
```
Comment (lispc) [inline zkevm-circuits/src/tx_circuit.rs:2642]: `pows_of_rand.push(pows_of_rand.last().unwrap().mul(keccak_input));` @kunxian-xia  we need to double check this line.   It is usually a dangerous pattern to unwrap challenge Value

File: zkevm-circuits/src/tx_circuit.rs [hunk 3]
```diff
@@ -3444,7 +3740,22 @@ impl<F: Field> TxCircuit<F> {
...
+                }
+
+                let chunk_txbytes_hash = keccak256(chunk_bytes.as_slice());
+                let evm_word = challenges.evm_word();
+                let chunk_txbytes_hash = rlc_be_bytes(&chunk_txbytes_hash, evm_word);
+
                 let mut tx_value_cells = vec![];
+                let mut chunk_txbytes_rlc_acc = Value::known(F::zero());
+                let mut chunk_txbytes_len_acc = Value::known(F::zero());
+                let mut pows_of_rand: Vec<Value<F>> = vec![Value::known(F::one())];
```
Comment (lispc) [inline zkevm-circuits/src/tx_circuit.rs:3786]: is this ok? `let mut pows_of_rand: Vec<Value<F>> = vec![Value::known(F::one())];`

File: zkevm-circuits/src/tx_circuit.rs [hunk 4]
```diff
@@ -3444,7 +3740,22 @@ impl<F: Field> TxCircuit<F> {
...
+                let mut chunk_bytes: Vec<u8> = vec![];
+                for i in 0..sigs.len() {
+                    let tx = get_tx(i);
+                    if (tx.tx_type != TxType::L1Msg) && !tx.caller_address.is_zero() {
+                        chunk_bytes.extend_from_slice(&tx.rlp_signed);
+                    }
+                }
+
+                let chunk_txbytes_hash = keccak256(chunk_bytes.as_slice());
+                let evm_word = challenges.evm_word();
```
Comment (lispc) [inline zkevm-circuits/src/tx_circuit.rs:3780]: can we use `challenges.keccak_challenge();` here?

File: zkevm-circuits/src/tx_circuit/test.rs [hunk 5]
```diff
@@ -146,7 +146,12 @@ fn build_eip1559_tx(id: usize) -> Transaction {
     tx.max_priority_fee_per_gas = eth_tx.max_priority_fee_per_gas.unwrap_or(U256::zero());
     tx.call_data = eth_tx.input.to_vec();
     tx.callee_address = eth_tx.to;
-    tx.caller_address = eth_tx.from;
+
+    tx.caller_address = H160::from_slice(&[
+        0x20, 0x2b, 0xB2, 0xfa, 0xB1, 0xe3, 0x5D, 0x94, 0x0F, 0xde, 0x99, 0xb2, 0x14, 0xba, 0x49,
+        0xDA, 0xfb, 0xce, 0xf6, 0x2A,
+    ]);
```
Comment (lispc) [inline zkevm-circuits/src/tx_circuit/test.rs:153]: @darth-cy what happened here?

**Comment by lispc** [comment]
@kunxian-xia need we change num_rows(ccc) methods of pi / tx circuits?

File: zkevm-circuits/src/pi_circuit.rs (hunk 1 above)
Comment (lispc) [reply zkevm-circuits/src/pi_circuit.rs]: i think the old codes are strictly more expressive.  Would revert

File: zkevm-circuits/src/tx_circuit/test.rs (hunk 5 above)
Comment (darth-cy) [reply zkevm-circuits/src/tx_circuit/test.rs:153]: The invocation on line 127 `EthTransaction::decode` failed to decode rlp bytes correctly and always produces a tx with caller address `0x000...000`. The resulting 1559 tx then gets treated as padding because of the 0x0 caller address. Since this is a method in a dependency, the only option is to manually assign the correct caller address.

Comment (lispc) [reply zkevm-circuits/src/tx_circuit/test.rs:153]: Cc @DreamWuGit similar problem here. Maybe we have issues with signing 1559

**Comment by roynalnaruto** [comment]
@darth-cy May be in the `PiCircuit` we should replace `max_txs` field by two separate fields `max_l1_msg_txs` and `max_l2_txs`. Currently we use `max_txs` to populate (and pad) both sections for L1 msg txs and L2 chunk bytes as well. Wdyt?

File: zkevm-circuits/src/pi_circuit.rs [hunk 6]
```diff
@@ -1015,9 +1202,260 @@ impl<F: Field> PiCircuitConfig<F> {
...
 
+    /// Assign chunk txbytes, the pre-image to chunk_txbytes_hash
+    /// i.e. keccak256(rlc(chunk_txbytes)) == chunk_txbytes_hash.
+    ///
+    /// The layout of the chunk_txbytes_hash section of the PI circuit differs slightly from other
+    /// sections. Instead of one byte per row, each row copies the TxHashRLC field from
+    /// TxCircuit, which is an RLC summary of rlp_signed (including tx_type byte) of a L2 tx.
+    /// This is to ensure the fixed length of an otherwise variable length section.
+    ///
+    /// The rip_length_acc field, subsequently, accumulates the length of the rlp_signed data for
```
Comment (roynalnaruto) [inline zkevm-circuits/src/pi_circuit.rs]:
```suggestion
    /// The rpi_length_acc field, subsequently, accumulates the length of the rlp_signed data for
```

File: zkevm-circuits/src/pi_circuit.rs [hunk 7]
```diff
@@ -518,6 +570,32 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
             },
         );
 
+        // 2b. constrain rpi_rlc and rpi_length_acc within chunk_txbytes section
+        meta.create_gate(
```
Comment (roynalnaruto) [inline zkevm-circuits/src/pi_circuit.rs]: You also need to constrain `rpi_length_acc`.

For non-padded rows, the lookup to the TxTable handles it, but you need to check that for padded rows, the `rpi_length_acc` does not change.

You might also need to check that the `rpi_length_acc` at the _first row_ of the `q_chunk_txbytes` section is correct? The lookup checks the length differential, so I believe you must be sure that the `rpi_length_acc` on the previous row is `0` (but may be this is already true because of the `init` row). Please check though

**Review by roynalnaruto — DISMISSED** [review]
pi circuit (first pass)

File: zkevm-circuits/src/tx_circuit.rs (hunk 2 above)
Comment (kunxian-xia) [reply zkevm-circuits/src/tx_circuit.rs:2642]: This unwrap is applied to `pows_of_rand.last()` which is of type `Option<&Value<F>>` **not** to `Value<F>`.

File: zkevm-circuits/src/pi_circuit.rs [hunk 8]
```diff
@@ -194,7 +200,27 @@ impl PublicData {
...
 
-    fn pi_bytes(&self, data_hash: H256) -> Vec<u8> {
+    /// Obtain the l2 tx (not padding; right now padding txs are l2 txs by default) bytes in the
+    /// chunk
+    fn chunk_txbytes(&self) -> Vec<u8> {
+        let mut result: Vec<u8> = vec![];
+        let chunk_txs_iter = self
+            .transactions
+            .iter()
+            .filter(|&tx| tx.tx_type != TxType::L1Msg && !tx.caller_address.is_zero());
```
Comment (kunxian-xia) [inline zkevm-circuits/src/pi_circuit.rs]: how about add a new method `is_padding` and put this line into it as this will be easier to understand?

File: zkevm-circuits/src/pi_circuit.rs (hunk 7 above)
Comment (darth-cy) [reply zkevm-circuits/src/pi_circuit.rs]: `rpi_length_acc` consistency constraint for padding rows is added in this patch: https://github.com/scroll-tech/zkevm-circuits/pull/1180. 

The correctness of `rpi_length_acc` at the first row is indeed included by the TxTable lookup as the `chunk_txbytes` section also has the `init` row as a padding with `rpi_length_acc` being 0.

**Comment by darth-cy** [comment]
> @darth-cy May be in the `PiCircuit` we should replace `max_txs` field by two separate fields `max_l1_msg_txs` and `max_l2_txs`. Currently we use `max_txs` to populate (and pad) both sections for L1 msg txs and L2 chunk bytes as well. Wdyt?

Perhaps having one maximum capacity parameter will work better? If the maximum capacity is split in two based strictly on tx type, then the network won't be able to shift capacity if one side becomes congested. I'm assuming for the newly defined `max_l1_msg_txs` and `max_l2_tx`, it will be `max_l1_msg_txs` + `max_l2_tx` <= `max_txs` (the current single maximum capacity).

**Comment by kunxian-xia** [comment]
> @kunxian-xia need we change num_rows(ccc) methods of pi / tx circuits?

@darth-cy I think we need to make sure that block's l2 tx bytes must not exceed the hard limit `4096*31 - 62` in `min_num_rows_blocks` (aka. ccc of tx circuit) which is exposed by batch prover .

File: zkevm-circuits/src/tx_circuit.rs [hunk 9]
```diff
@@ -1690,6 +1731,156 @@ impl<F: Field> SubCircuitConfig<F> for TxCircuitConfig<F> {
...
+                    not::expr(meta.query_advice(is_chunk_bytes, Rotation::cur())),
+                    is_hash_rlc(meta),
+                ]))
+            },
+        );
+
+        meta.lookup_any("Correct pow_of_rand for HashLen", |meta| {
+            let enable = and::expr(vec![
+                meta.query_fixed(q_enable, Rotation::cur()),
+                not::expr(meta.query_advice(is_l1_msg, Rotation::cur())),
```
Comment (kunxian-xia) [inline zkevm-circuits/src/tx_circuit.rs:1864]: The condition should be `q_enable && is_chunk_bytes && is_hash_rlc` as we use `pow_of_rand::cur` to accumulate `chunk_txbytes_rlc` at `HASH_RLC` row.

File: zkevm-circuits/src/tx_circuit.rs [hunk 10]
```diff
@@ -2367,7 +2600,30 @@ impl<F: Field> TxCircuitConfig<F> {
         let (access_list_address_size, access_list_storage_key_size) =
             access_list_size(&tx.access_list);
 
-        // fixed_rows of a tx
+        // Only bytes from L2 txs are accumulated for chunk bytes hash
+        let is_chunk_bytes = (tx.tx_type != TxType::L1Msg) && !tx.caller_address.is_zero();
```
Comment (kunxian-xia) [inline zkevm-circuits/src/tx_circuit.rs]: It's better to add a new function `is_real_l2_tx` for type `Transaction` and put these two conditions into it.

File: zkevm-circuits/src/tx_circuit.rs [hunk 11]
```diff
@@ -3309,6 +3596,15 @@ impl<F: Field> TxCircuit<F> {
         let sign_verify_inputs = keccak_inputs_sign_verify(&sign_datas);
         inputs.extend_from_slice(&sign_verify_inputs);
 
+        // Keccak input for chunk bytes (only L2 txs are included)
+        let chunk_hash_bytes = self
+            .txs
+            .iter()
+            .filter(|&tx| (tx.tx_type != TxType::L1Msg) && !tx.caller_address.is_zero())
```
Comment (kunxian-xia) [inline zkevm-circuits/src/tx_circuit.rs]: how about using `tx.is_real_l2_tx()`?

File: zkevm-circuits/src/tx_circuit.rs (hunk 4 above)
Comment (kunxian-xia) [reply zkevm-circuits/src/tx_circuit.rs:3780]: No, because we are storing keccak digest's result.

File: zkevm-circuits/src/tx_circuit.rs (hunk 3 above)
Comment (kunxian-xia) [reply zkevm-circuits/src/tx_circuit.rs:3786]: yes, as the current max real l2 tx bytes' len is 0.

File: zkevm-circuits/src/tx_circuit.rs [hunk 12]
```diff
@@ -330,6 +353,20 @@ impl<F: Field> SubCircuitConfig<F> for TxCircuitConfig<F> {
...
+        // Chunk bytes accumulator
+        let is_chunk_bytes = meta.advice_column();
+        let chunk_bytes_len = meta.advice_column();
+        let chunk_txbytes_rlc = meta.advice_column_in(SecondPhase);
+        let chunk_txbytes_hash = meta.advice_column_in(SecondPhase);
+        let chunk_txbytes_len_acc = meta.advice_column();
+        let pow_of_rand = meta.advice_column_in(SecondPhase);
+
+        meta.enable_equality(chunk_bytes_len);
+        meta.enable_equality(chunk_txbytes_rlc);
```
Comment (kunxian-xia) [inline zkevm-circuits/src/tx_circuit.rs:372]: why do we need to put these 4 columns except `chunk_txbytes_hash` into permutation argument? It seems only the last tx's `chunk_txbytes_hash` cell needs to be copied to pi circuit.

**Comment by lispc** [comment]
fixed.    cc.  @roynalnaruto see my last comit.  <del>i will temporarily always use pow of rand table len == 4094 * 31.  Seems relying on cfg!(feature = "test") does now work well.  A lot of codes rely on that feature, it is not easy to disable it.   (btw aggregator crate enables zkevm-circuit test feature)</del>

File: zkevm-circuits/src/evm_circuit/execution/end_inner_block.rs [hunk 13]
```diff
@@ -98,7 +98,7 @@ impl<F: Field> ExecutionGadget<F> for EndInnerBlockGadget<F> {
             rw_counter: Transition::Same,
             // We propagate call_id so that EndBlock can get the last tx_id
             // in order to count processed txs.
-            call_id: Transition::Same,
```
Comment (z2trillion) [inline zkevm-circuits/src/evm_circuit/execution/end_inner_block.rs:101]: Is this removal intentional?

Comment (lispc) [reply zkevm-circuits/src/evm_circuit/execution/end_inner_block.rs:101]: Yes. It is wrong.

**Comment by lispc** [comment]
confirmed: the last commit [4f5c3bb](https://github.com/scroll-tech/zkevm-circuits/pull/1153/commits/4f5c3bbfbc7309335cce13d50626d41b3d9c63ed) does not affect vk.

File: zkevm-circuits/src/tx_circuit.rs [hunk 14]
```diff
@@ -3600,17 +3950,62 @@ impl<F: Field> SubCircuit<F> for TxCircuit<F> {
...
         };
-        let tx_usage = sum_calldata_len as f32 / max_calldata as f32;
+        let dynamic_usage =
+            (sum_calldata_len + sum_access_list_len) as f32 / max_dynamic_data as f32;
+
+        // Get the highest usage fraction out of all capacities
+        let highest_usage = ([blob_usage, dynamic_usage])
+            .iter()
+            .cloned()
+            .fold(0_f32, f32::max);
```
Comment (lispc) [inline zkevm-circuits/src/tx_circuit.rs:4003]: can be simplified...?  @darth-cy

File: zkevm-circuits/src/tx_circuit/test.rs (hunk 5 above)
Comment (lispc) [reply zkevm-circuits/src/tx_circuit/test.rs:153]: @darth-cy https://github.com/scroll-tech/zkevm-circuits/pull/1182 merged, you can have a look whether you still need this adhoc change. If yes,  add a "FIXME" comment.

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -38,6 +38,7 @@ jobs:
 
   test-default:
     name: Test Default
+    if: false
```
Comment (lispc) [inline .github/workflows/ci.yml]: CI seems not work well now. So L1 mode zkEVM tests are disabled.   FIXME later.

**Review by lispc — APPROVED** [review]
